File: codex_opencode_adapter/conversion/tool_context.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

from .ids import compact_json

JSON = dict[str, Any]
_SAFE_TOOL_NAME = re.compile(r"[^a-zA-Z0-9_-]")
CHAT_TOOL_NAME_MAX_LEN = 64
CUSTOM_TOOL_INPUT_FIELD = "input"


@dataclass
class ToolSpec:
    kind: str
    name: str
    chat_name: str
    namespace: str | None = None


@dataclass
class ToolContext:
    chat_tools: list[JSON] = field(default_factory=list)
    reverse_names: dict[str, str] = field(default_factory=dict)
    specs_by_chat_name: dict[str, ToolSpec] = field(default_factory=dict)
    namespace_lookup: dict[tuple[str, str], str] = field(default_factory=dict)
    _used: set[str] = field(default_factory=set)

    def is_custom_tool_chat_name(self, name: str) -> bool:
        spec = self.specs_by_chat_name.get(name)
        return bool(spec and spec.kind in {"custom", "tool_search"})

    def restore_name(self, name: str) -> str:
        return self.reverse_names.get(name, name)

    def add_response_tool(self, item: Any) -> None:
        if isinstance(item, str):
            self._add_custom({"type": "custom", "name": item})
            return
        if not isinstance(item, dict):
            return
        kind = str(item.get("type") or "function")
        if kind == "namespace":
            self._add_namespace(item)
        elif kind in {"custom", "tool_search"}:
            self._add_custom(item, kind=kind)
        else:
            source = _tool_source(item)
            if isinstance(source, dict):
                self._add_function(source)

    def _safe_name(self, original: str) -> str:
        base = _SAFE_TOOL_NAME.sub("_", original.strip())[:CHAT_TOOL_NAME_MAX_LEN] or "tool"
        candidate = base
        suffix = 2
        while candidate in self._used:
            tail = f"_{suffix}"
            candidate = f"{base[: CHAT_TOOL_NAME_MAX_LEN - len(tail)]}{tail}"
            suffix += 1
        self._used.add(candidate)
        return candidate

    def _add_chat_tool(self, original: str, spec: ToolSpec, parameters: JSON, description: str) -> None:
        if not original:
            return
        self.reverse_names[spec.chat_name] = original
        self.specs_by_chat_name[spec.chat_name] = spec
        if spec.namespace:
            self.namespace_lookup[(spec.namespace, spec.name)] = spec.chat_name
        self.chat_tools.append(
            {
                "type": "function",
                "function": {
                    "name": spec.chat_name,
                    "description": description,
                    "parameters": parameters,
                },
            }
        )

    def _add_function(self, source: JSON, *, original_name: str | None = None, namespace: str | None = None) -> None:
        original = str(original_name or source.get("name") or "").strip()
        if not original:
            return
        chat_name = self._safe_name(original)
        parameters = source.get("parameters")
        if not isinstance(parameters, dict):
            parameters = {"type": "object", "properties": {}}
        spec = ToolSpec(
            kind="namespace" if namespace else "function",
            name=str(source.get("name") or original),
            chat_name=chat_name,
            namespace=namespace,
        )
        self._add_chat_tool(original, spec, parameters, str(source.get("description") or ""))

    def _add_custom(self, source: JSON, *, kind: str = "custom") -> None:
        original = str(source.get("name") or "").strip()
        if not original:
            return
        description = str(source.get("description") or "")
        preserved = compact_json(source)
        desc = (
            f"{description}\n\nOriginal Responses custom tool definition:\n{preserved}"
            if description
            else f"Original Responses custom tool definition:\n{preserved}"
        )
        chat_name = self._safe_name(original)
        spec = ToolSpec(kind=kind, name=original, chat_name=chat_name)
        self._add_chat_tool(
            original,
            spec,
            {
                "type": "object",
                "properties": {
                    CUSTOM_TOOL_INPUT_FIELD: {
                        "type": "string",
                        "description": "Raw string input for the original custom tool.",
                    }
                },
                "required": [CUSTOM_TOOL_INPUT_FIELD],
            },
            desc,
        )

    def _add_namespace(self, source: JSON) -> None:
        namespace = str(source.get("name") or "").strip()
        children = source.get("tools") or source.get("children") or []
        if not namespace or not isinstance(children, list):
            return
        for child in children:
            if not isinstance(child, dict):
                continue
            child_source = _tool_source(child)
            if not isinstance(child_source, dict):
                continue
            child_name = str(child_source.get("name") or "").strip()
            if child_name:
                self._add_function(child_source, original_name=f"{namespace}__{child_name}", namespace=namespace)


def _tool_source(item: JSON) -> JSON | None:
    nested = item.get("function")
    return nested if isinstance(nested, dict) else item


def build_tool_context(tools: Any) -> ToolContext:
    context = ToolContext()
    if isinstance(tools, list):
        for item in tools:
            context.add_response_tool(item)
    return context


def convert_tools(tools: Any) -> tuple[list[JSON], dict[str, str]]:
    context = build_tool_context(tools)
    return context.chat_tools, context.reverse_names


def restore_tool_name(name: str, reverse: dict[str, str]) -> str:
    return reverse.get(name, name)


def convert_tool_choice(tool_choice: Any, context: ToolContext) -> Any:
    if tool_choice is None:
        return None
    if isinstance(tool_choice, str):
        return tool_choice
    if not isinstance(tool_choice, dict):
        return tool_choice
    kind = tool_choice.get("type")
    if kind in {"auto", "none", "required"}:
        return kind
    if kind in {"function", "tool"}:
        name = str(tool_choice.get("name") or "")
        chat_name = next((safe for safe, original in context.reverse_names.items() if original == name), name)
        return {"type": "function", "function": {"name": chat_name}}
    return tool_choice

File: src/conversion/tool_context.rs
use serde_json::{json, Value};
use std::collections::{HashMap, HashSet};

const CHAT_TOOL_NAME_MAX_LEN: usize = 64;
const CUSTOM_TOOL_INPUT_FIELD: &str = "input";
const TOOL_SEARCH_NAME: &str = "tool_search";

pub fn custom_tool_input_field() -> &'static str {
    CUSTOM_TOOL_INPUT_FIELD
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ToolKind {
    Function,
    Namespace,
    Custom,
    ToolSearch,
}

#[derive(Debug, Clone)]
pub struct ToolSpec {
    pub kind: ToolKind,
    pub name: String,
    pub chat_name: String,
    pub namespace: Option<String>,
}

#[derive(Debug, Default, Clone)]
pub struct ToolContext {
    pub chat_tools: Vec<Value>,
    pub reverse_names: HashMap<String, String>,
    specs_by_chat_name: HashMap<String, ToolSpec>,
    namespace_name_to_chat_name: HashMap<(String, String), String>,
    used: HashSet<String>,
}

impl ToolContext {
    pub fn build(tools: Option<&Value>) -> Self {
        Self::build_with_input(tools, None)
    }

    pub fn build_with_input(tools: Option<&Value>, input: Option<&Value>) -> Self {
        let mut context = Self::default();
        if let Some(Value::Array(items)) = tools {
            for item in items {
                context.add_response_tool(item);
            }
        }
        if let Some(input) = input {
            context.collect_tool_search_output_tools(input);
        }
        context
    }

    pub fn restore_name(&self, name: &str) -> String {
        self.reverse_names.get(name).cloned().unwrap_or_else(|| name.to_string())
    }

    pub fn is_custom_tool_chat_name(&self, name: &str) -> bool {
        self.specs_by_chat_name
            .get(name)
            .map(|spec| matches!(spec.kind, ToolKind::Custom))
            .unwrap_or(false)
    }

    pub fn is_tool_search_chat_name(&self, name: &str) -> bool {
        self.specs_by_chat_name
            .get(name)
            .map(|spec| matches!(spec.kind, ToolKind::ToolSearch))
            .unwrap_or(false)
    }

    pub fn lookup_spec(&self, chat_name: &str) -> Option<&ToolSpec> {
        self.specs_by_chat_name.get(chat_name)
    }

    pub fn chat_name_for_response_function(&self, name: &str, namespace: Option<&str>) -> String {
        if let Some(namespace) = namespace.filter(|value| !value.is_empty()) {
            if let Some(chat_name) = self.namespace_name_to_chat_name.get(&(namespace.to_string(), name.to_string())) {
                return chat_name.clone();
            }
            return format!("{namespace}__{name}").chars().take(CHAT_TOOL_NAME_MAX_LEN).collect();
        }
        self.reverse_names
            .iter()
            .find_map(|(safe, original)| (original == name).then_some(safe.clone()))
            .unwrap_or_else(|| name.to_string())
    }

    fn add_response_tool(&mut self, item: &Value) {
        if let Some(name) = item.as_str() {
            self.add_custom(&json!({"type":"custom","name":name}), ToolKind::Custom);
            return;
        }
        let Some(obj) = item.as_object() else { return; };
        match obj.get("type").and_then(Value::as_str).unwrap_or("function") {
            "namespace" => self.add_namespace(item),
            "custom" => self.add_custom(item, ToolKind::Custom),
            "tool_search" => self.add_tool_search(item),
            _ => self.add_function(tool_source(item).unwrap_or(item), None, None),
        }
    }

    fn add_namespace(&mut self, item: &Value) {
        let namespace = item.get("name").and_then(Value::as_str).unwrap_or("").trim();
        let Some(children) = item.get("tools").or_else(|| item.get("children")).and_then(Value::as_array) else { return; };
        if namespace.is_empty() { return; }
        for child in children {
            let source = tool_source(child).unwrap_or(child);
            if let Some(child_name) = source.get("name").and_then(Value::as_str).filter(|s| !s.trim().is_empty()) {
                self.add_function(source, Some(format!("{namespace}__{child_name}")), Some(namespace.to_string()));
            }
        }
    }

    fn add_custom(&mut self, item: &Value, kind: ToolKind) {
        let original = item.get("name").and_then(Value::as_str).unwrap_or("").trim();
        if original.is_empty() { return; }
        let chat_name = self.safe_name(original);
        let description = item.get("description").and_then(Value::as_str).unwrap_or("");
        let preserved = serde_json::to_string(item).unwrap_or_default();
        let description = if description.is_empty() {
            format!("Original Responses custom tool definition:\n{preserved}")
        } else {
            format!("{description}\n\nOriginal Responses custom tool definition:\n{preserved}")
        };
        let spec = ToolSpec { kind, name: original.to_string(), chat_name: chat_name.clone(), namespace: None };
        self.add_chat_tool(original, spec, json!({
            "type": "object",
            "properties": {
                CUSTOM_TOOL_INPUT_FIELD: {
                    "type": "string",
                    "description": "Raw string input for the original custom tool."
                }
            },
            "required": [CUSTOM_TOOL_INPUT_FIELD]
        }), description);
    }

    fn add_tool_search(&mut self, item: &Value) {
        let original = TOOL_SEARCH_NAME;
        let chat_name = self.safe_name(original);
        let description = item
            .get("description")
            .and_then(Value::as_str)
            .unwrap_or("Search for tools available to the client.")
            .to_string();
        let spec = ToolSpec {
            kind: ToolKind::ToolSearch,
            name: original.to_string(),
            chat_name: chat_name.clone(),
            namespace: None,
        };
        self.add_chat_tool(original, spec, json!({
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Search query for tools."
                },
                "limit": {
                    "type": "integer",
                    "description": "Maximum number of tools to return."
                }
            },
            "required": ["query"]
        }), description);
    }

    fn add_function(&mut self, source: &Value, original_name: Option<String>, namespace: Option<String>) {
        let original = original_name.unwrap_or_else(|| source.get("name").and_then(Value::as_str).unwrap_or("").trim().to_string());
        if original.is_empty() { return; }
        let chat_name = self.safe_name(&original);
        let parameters = source.get("parameters").cloned().filter(Value::is_object).unwrap_or_else(|| json!({"type":"object","properties":{}}));
        let spec = ToolSpec {
            kind: if namespace.is_some() { ToolKind::Namespace } else { ToolKind::Function },
            name: source.get("name").and_then(Value::as_str).unwrap_or(&original).to_string(),
            chat_name: chat_name.clone(),
            namespace: namespace.clone(),
        };
        if let Some(namespace) = namespace {
            self.namespace_name_to_chat_name.insert((namespace, spec.name.clone()), chat_name.clone());
        }
        self.add_chat_tool(original.as_str(), spec, parameters, source.get("description").and_then(Value::as_str).unwrap_or("").to_string());
    }

    fn add_chat_tool(&mut self, original: &str, spec: ToolSpec, parameters: Value, description: String) {
        self.reverse_names.insert(spec.chat_name.clone(), original.to_string());
        self.specs_by_chat_name.insert(spec.chat_name.clone(), spec.clone());
        self.chat_tools.push(json!({
            "type": "function",
            "function": {
                "name": spec.chat_name,
                "description": description,
                "parameters": parameters,
            }
        }));
    }

    fn collect_tool_search_output_tools(&mut self, value: &Value) {
        match value {
            Value::Array(items) => {
                for item in items {
                    self.collect_tool_search_output_tools(item);
                }
            }
            Value::Object(obj) => {
                if obj.get("type").and_then(Value::as_str) == Some("tool_search_output") {
                    if let Some(tools) = obj.get("tools").and_then(Value::as_array) {
                        for tool in tools {
                            self.add_response_tool(tool);
                        }
                    }
                }
                for value in obj.values() {
                    self.collect_tool_search_output_tools(value);
                }
            }
            _ => {}
        }
    }

    fn safe_name(&mut self, original: &str) -> String {
        let mut base = original
            .chars()
            .map(|c| if c.is_ascii_alphanumeric() || c == '_' || c == '-' { c } else { '_' })
            .take(CHAT_TOOL_NAME_MAX_LEN)
            .collect::<String>();
        if base.is_empty() { base = "tool".to_string(); }
        let mut candidate = base.clone();
        let mut suffix = 2;
        while self.used.contains(&candidate) {
            let tail = format!("_{suffix}");
            let keep = CHAT_TOOL_NAME_MAX_LEN.saturating_sub(tail.len());
            let head = base.chars().take(keep).collect::<String>();
            candidate = format!("{head}{tail}");
            suffix += 1;
        }
        self.used.insert(candidate.clone());
        candidate
    }
}

fn tool_source(item: &Value) -> Option<&Value> {
    item.get("function").filter(|v| v.is_object()).or(Some(item))
}
